spatial-tracking-explainer.md

```diff
@@ -0,0 +1,382 @@
+| `XRStationaryFrameOfReference` | `eye-level`         | Racing game or space simulator                |
+| `XRStationaryFrameOfReference` | `floor-level`       | Action game where you duck and dodge in place |
+| `XRBoundedFrameOfReference`    |                     | Room redecorator                              |
+| `XRUnboundedFrameOfReference`  |                     | Walking tour                                  |
+
+### XRFrameOfReference Availability
+
+**Guaranteed** The UA will always be able to provide this frame of reference 
+
+**Hardware-dependent** The UA will only be able to supply this frame of reference if running on XR hardware that supports it
+
+**Rejected** The UA will never provide this frame of reference
+
+&ast; Or polyfillable on inside-out tracking systems with anchor support.  See the section on "Floor Alignment Compatibility".
+
+| Type                           | Subtype             | Inline             | Immersive |
+| -------------------------------| ------------------- | ------------------ | ---------- |
+| `XRStationaryFrameOfReference` | `position-disabled` | Hardware-dependent | Guaranteed |
+| `XRStationaryFrameOfReference` | `eye-level`         | Hardware-dependent | Guaranteed |
+| `XRStationaryFrameOfReference` | `floor-level`       | Hardware-dependent | Guaranteed* |
+| `XRBoundedFrameOfReference`    |                     | Rejected           | Hardware-dependent* |
+| `XRUnboundedFrameOfReference`  |                     | Rejected           | Hardware-dependent |
+
+## Appendix B: Proposed partial IDL
+This is a partial IDL and is considered additive to the core IDL found in the main [explainer](explainer.md).
+
+```webidl
+//
+// Session
+//
+
+partial dictionary XRSessionCreationOptions {
+  XRFrameOfReferenceType requiredFrameOfReferenceType;
+};
+
+partial interface XRSession {
+  Promise<XRFrameOfReference> requestFrameOfReference(XRFrameOfReferenceOptions options);
+};
+
+//
+// Frames and Poses
+//
+
+partial interface XRFrame {
+  // Also listed in the main explainer.md
+  XRDevicePose? getDevicePose(optional XRFrameOfReference frameOfReference);
+  XRInputPose? getInputPose(XRInputSource inputSource, optional XRFrameOfReference frameOfReference);
+};
+
+[SecureContext, Exposed=Window]
+interface XRInputPose {
+  readonly attribute boolean emulatedPosition;
+  readonly attribute XRRay targetRay;
+  readonly attribute Float32Array? gripMatrix;
+};
+
+//
+// Coordinate System
+//
+
+[SecureContext, Exposed=Window] interface XRCoordinateSystem {
+  Float32Array? getTransformTo(XRCoordinateSystem other);
+};
+
+//
+// Frame of Reference
+//
+
+enum XRFrameOfReferenceType {
+  "stationary",
+  "bounded",
+  "unbounded"
+};
+
+dictionary XRFrameOfReferenceOptions {
+  XRFrameOfReferenceType type;
+};
+
+[SecureContext, Exposed=Window] interface XRFrameOfReference : EventTarget {
+  readonly attribute XRCoordinateSystem coordinateSystem;
```

Comment:
I don't see why we need this - what would be an example use-case? My belief is that the only thing that we need to get out of XRFrameOfReference in the end is a view matrix. It ultimately represents the way that the device pose is being represented and calculated, which, when used to generate a camera in a virtual world, is not needed beyond that since the virtual objects in the world can exist in a simple substrate reference frame.

If the only reason for this added complexity is disconnected spaces in AR where we don't know how they relate to each other and need a way of knowing whether/how some system of objects/space should be rendered, then it seems worth exploring that as part of this explainer. As it stands, I don't see a lot of justification for this complexity and would like it to be described as a companion to this IDL.

Reply:
If you can locate a plane/mesh/etc.'s `XRCoordinateSystem` relative to your base frame at the moment, you can use that model transform to render the object this frame.  If not, you should hide the object this frame.

You're correct in noting that some of the concepts in the doc (`"unbounded"` frame and `XRCoordinateSystem`) are poorly motivated at the moment, as we have yet to introduce multiple independent anchors to the API.  Some of this complexity is scaffolding for those future APIs.

Reply:
@thetuvix I think Max's question is related to the use of a single `XRUnboundedFrameOfReference`.  When you're in room A and move to room B by going down a long hallway... what is your transform when you get tracking back in room B.

Reply:
> When you're in room A and move to room B by going down a long hallway... what is your transform when you get tracking back in room B.

There is a separate follow-up here where we need to decide for WebXR what common tracking loss/recovery behavior to establish as a developer promise for the various frames

At least in Windows MR, when you lose tracking, the WinMR equivalent of an `XRUnboundedFrameOfReference` will no longer provide you a device position.  When you regain tracking:
* If the system finds you somewhere back within the previous map fragment, your device position in the `XRUnboundedFrameOfReference` will jump there, keeping the origin in place.
* If the system _doesn't_ find you somewhere within the previous map fragment, your device position will stay at the same XYZ position in the `XRUnboundedFrameOfReference`, which has the effect of having "dragged" the unbounded origin along with you during the period of tracking loss.

Regardless, when you are in either map fragment, if you ask for a transform from a free-space anchor in the other fragment to your unbounded frame, you'll get back null.

Reply:
Do you happen to have a table with the tracking loss behavior for many of the "major" systems?  

Also, as we dig into the handling of tracking loss, perhaps this is a case for adding the `onreset` event to the `XRUnboundedFrameOfReference`?

Reply:
Does this mean you have to know that you need to request a new frame-of-reference when you get tracking back? Is there an event-driven API for this? I know Nell said she isn't covering tracking loss in this explainer, so maybe this detail doesn't belong here but there is definitely some overlap with respect to how coordinate systems are used to render virtual objects. It is still not clear to me how an app would use these APIs to deal with this situation.

Reply:
Yeah, reading over all the comments this morning... I think we probably do need to address tracking loss before trying to merge this PR.  I'll dig into it

Reply:
@lincolnfrog:
> Does this mean you have to know that you need to request a new frame-of-reference when you get tracking back?

No - in Windows MR, the one unbounded frame of reference object you created at app startup continues to work after tracking loss, with the behavior described by the bullets above.  I've been presuming the same approach in WebXR.

We can perhaps separate two things here:
* Documenting the fact that all `XRFrameOfReference` instances are durable objects across tracking loss and recovery.
* Nailing down the precise details of that tracking loss/recovery behavior, such as the positions and orientations you get back after tracking recovery in various situations.

We could perhaps update this PR to document the former, while iterating on the latter in a follow-on PR.

Reply:
I've moved the `onreset` to the `XRFrameOfReference` base interface, but I still think the broader topic of tracking loss may need to be addressed before we can credibly merge this PR.

Reply:
@thetuvix how does the app know what virtual objects are no longer valid (i.e. belong to the old map fragment)?

Reply:
@lincolnfrog 

Each frame, for any given anchor:
* If a pose **is available** for that anchor in your frame of reference, you should render the anchor's content there.
* If a pose **is not available** for that anchor in your base frame of reference, you should hide the anchor's content until a pose is available again.

Reply:
There are definitely questions about anchor culling that we'll need to take into consideration, but that should be part of a different PR ^_^

Reply:
This seems like we've come to a good conclusion, and the next iteration will have (hopefully) addressed this adequately.  Please reopen if that turns out not to be the case.